ambry-store/src/main/java/com.github.ambry.store/LogSegment.java

```diff
@@ -0,0 +1,339 @@
+   * {@code size} data.
+   * @throws IOException if data could not be written to the file because of I/O errors
+   *
+   */
+  void writeFrom(ReadableByteChannel channel, long offset, long size)
+      throws IOException {
+    if (offset < 0 || offset > capacityInBytes) {
+      throw new IllegalArgumentException(
+          "Provided offset [" + offset + "] is out of bounds for the segment [" + file.getAbsolutePath()
+              + "] with capacity [" + capacityInBytes + "]");
+    }
+    if (offset + size > capacityInBytes) {
+      metrics.overflowWriteError.inc();
+      throw new IllegalArgumentException("Cannot write to segment [" + file.getAbsolutePath() + "] from offset [" +
+          offset + "] for size [" + size + "] because it exceeds the capacity [" + capacityInBytes + "]");
+    }
+    long bytesWritten = 0;
+    while (bytesWritten < size) {
+      bytesWritten += fileChannel.transferFrom(channel, offset + bytesWritten, size - bytesWritten);
+    }
+    if (offset + size > endOffset.get()) {
+      endOffset.set(offset + size);
+    }
+  }
+
+  /**
+   * Gets the {@link File} and {@link FileChannel} backing this log segment. Also increments a ref count.
+   * <p/>
+   * The expectation is that a matching {@link #closeView()} will be called eventually to decrement the ref count.
+   * @param offset the offset that will used to start the read operation from.
+   * @return the {@link File} and {@link FileChannel} backing this log segment.
+   * @throws IllegalArgumentException if {@code offset} < 0 or is > {@link #getEndOffset()}.
+   */
+  Pair<File, FileChannel> getView(long offset) {
+    if (offset < 0 || offset > getEndOffset()) {
+      throw new IllegalArgumentException(
+          "Provided offset [" + offset + "] is out of bounds for the segment [" + file.getAbsolutePath()
+              + "] with end offset [" + getEndOffset() + "]");
+    }
+    refCount.incrementAndGet();
+    return segmentView;
+  }
+
+  /**
+   * Closes view that was obtained (decrements ref count).
+   */
+  void closeView() {
+    refCount.decrementAndGet();
+  }
+
+  /**
+   * @return size of the backing file on disk.
+   * @throws IOException if the size could not be obtained due to I/O error.
+   */
+  long sizeInBytes()
+      throws IOException {
+    return fileChannel.size();
+  }
+
+  /**
+   * @return the name of this segment.
+   */
+  String getName() {
+    return name;
+  }
+
+  /**
+   * Sets the end offset of this segment. This can be lesser than the actual size of the file and represents the offset
+   * until which data that is readable is stored.
+   * @param endOffset the end offset of this log.
+   * @throws IllegalArgumentException if {@code endOffset} < 0 or {@code endOffset} > the size of the file.
+   * @throws IOException if there is any I/O error.
+   */
+  void setEndOffset(long endOffset)
+      throws IOException {
+    long fileSize = sizeInBytes();
+    if (endOffset < 0 || endOffset > fileSize) {
```

Comment:
It may create holes and it might push the end offset farther but as mentioned earlier, I expect the caller to know what is happening - I do not want to limit the flexibility of the function and tailor it to a single use case.